tests/NotifierTest.php

```diff
@@ -0,0 +1,155 @@
+				return \vsprintf($text, $parameters);
+			}));
+
+		$this->factory->method('get')->willReturn($l10n);
+
+		$this->notifier = new Notifier($this->factory, $this->timeFactory);
+	}
+
+	/**
+	 * @expectedException \InvalidArgumentException
+	 */
+	public function testPrepareInvalidApp() {
+		$notification = $this->createMock(INotification::class);
+		$notification->method('getApp')->willReturn('another');
+		$notification->method('getObjectType')->willReturn('local_share');
+		$this->notifier->prepare($notification, 'en_US');
+	}
+
+	/**
+	 * @expectedException \InvalidArgumentException
+	 */
+	public function testPrepareInvalidUnknownObjectType() {
+		$notification = $this->createMock(INotification::class);
+		$notification->method('getApp')->willReturn('password_policy');
+		$notification->method('getObjectType')->willReturn('local_share');
+		$this->notifier->prepare($notification, 'en_US');
+	}
+
+	public function testPrepareAboutToExpire() {
+		$initialTime = 1531232050;
+		$expireIn = 10 * 24 * 60 * 60;  // 10 days;
+		$notification = $this->createMock(INotification::class);
+		$notification->method('getApp')->willReturn('password_policy');
+		$notification->method('getObjectType')->willReturn('about_to_expire');
+		$notification->method('getSubjectParameters')->willReturn([$initialTime, $expireIn]);
+		// first parameter is the password change time and the second is the
+		// expiration time starting from the password change time
+		$notification->method('getMessageParameters')->willReturn([$initialTime, $expireIn]);
+
+		$this->timeFactory->method('getTime')->willReturn($initialTime + (7 * 24 * 60 * 60));
+
+		$notification->expects($this->once())
+			->method('setParsedSubject')
+			->with('Your password is about to expire!');
+
+		$notification->expects($this->once())
+			->method('setParsedMessage')
+			->with('You have 3 days to change your password');
+
+		$this->notifier->prepare($notification, 'en_US');
+	}
+
+	public function testPrepareAboutToExpirePassDate() {
+		$initialTime = 1531232050;
+		$expireIn = 10 * 24 * 60 * 60;  // 10 days;
+		$notification = $this->createMock(INotification::class);
+		$notification->method('getApp')->willReturn('password_policy');
+		$notification->method('getObjectType')->willReturn('about_to_expire');
+		$notification->method('getSubjectParameters')->willReturn([$initialTime, $expireIn]);
+		// first parameter is the password change time and the second is the
+		// expiration time starting from the password change time
+		$notification->method('getMessageParameters')->willReturn([$initialTime, $expireIn]);
+
+		$this->timeFactory->method('getTime')->willReturn($initialTime + (12 * 24 * 60 * 60));
+
+		$notification->expects($this->once())
+			->method('setParsedSubject')
+			->with('Your password is about to expire!');
+
+		$notification->expects($this->once())
+			->method('setParsedMessage')
+			->with('Your password has already expired 2 days ago');
```

Comment:
Hmmm - wherever in the real code this message is generated, the text could be more simple and straightforward:
``Your password expired 2 days ago``